singularity-eos/eos/eos_powermg.hpp

```diff
@@ -0,0 +1,440 @@
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_POWERMG_HPP_
+#define _SINGULARITY_EOS_EOS_POWERMG_HPP_
+
+#include <cmath>
+#include <cstdio>
+
+#include <limits>
+
+#include <ports-of-call/portable_errors.hpp>
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/math_utils.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+using namespace robust;
+
+// COMMENT: This is a complete, thermodynamically consistent Mie-Gruneisen
+// EOS based on a linear Us-up Hugoniot, Us=Cs+s*up.
+class PowerMG : public EosBase<PowerMG> {
+ public:
+  PowerMG() = default;
+  // Constructor
+  PowerMG(const Real rho0, const Real T0, const Real G0, const Real Cv0, const Real E0,
+          const Real S0, const Real Pmin, const Real *expconsts)
+      : _rho0(rho0), _T0(T0), _G0(G0), _Cv0(Cv0), _E0(E0), _S0(S0), _Pmin(Pmin) {
+    _InitializePowerMG(expconsts);
+    _CheckPowerMG();
+  }
+
+  PowerMG GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real
+  MinInternalEnergyFromDensity(const Real rho, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real EntropyFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real EntropyFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const;
+  PORTABLE_INLINE_FUNCTION Real SpecificHeatFromDensityTemperature(
+      const Real rho, const Real temp, Real *lambda = nullptr) const {
+    return _Cv0;
+  }
+  PORTABLE_INLINE_FUNCTION Real SpecificHeatFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return _Cv0;
+  }
+  // added for testing AEM Dec 2023
+  PORTABLE_INLINE_FUNCTION Real AllHugPressureFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real AllHugInternalEnergyFromDensity(const Real rho) const;
+  PORTABLE_INLINE_FUNCTION Real AllHugTemperatureFromDensity(const Real rho) const;
+  // Thermal Bulk Modulus added AEM Dec 2022
+  PORTABLE_INLINE_FUNCTION Real TBulkModulusFromDensityTemperature(
```

Comment:
No need to change this for now---but worth noting. If we ever expose thermal bulk mod, thermal expansion, etc., we may want to rename these. (Not on you. That's just me making a note for future reference.)